File: elixir/collatz.exs
defmodule Collatz do
  def collatz(seed) do
    collatz(seed, 0)
  end

  defp collatz(1, steps), do: steps

  defp collatz(seed, steps) when rem(seed, 2) == 0 do
    collatz(div(seed, 2), steps + 1)
  end

  defp collatz(seed, steps) do
    collatz(seed * 3 + 1, steps + 1)
  end
end

defmodule Main do
  def run do
    collatz_result = Enum.reduce(1..1_000_000, 0, fn seed, acc ->
      acc + Collatz.collatz(seed)
    end)
    IO.puts("steps #{collatz_result}")
  end
end

Main.run()

File: ruby/collatz.rb
def collatz(seed)
    steps = 0
    while seed > 1
      while seed % 2 == 0
        steps += 1
        seed /= 2
      end
      if seed > 1
        steps += 1
        seed = seed * 3 + 1
      end
    end
    steps
  end
  
  def main
    collatz_result = 0
    (1..1_000_000).each do |seed|
      collatz_result += collatz(seed)
    end
    puts "steps #{collatz_result}"
  end
  
  main
  